ddl/column.go

```diff
@@ -427,13 +427,106 @@ func (d *ddl) onSetDefaultValue(t *meta.Meta, job *model.Job) error {
 func (d *ddl) onModifyColumn(t *meta.Meta, job *model.Job) error {
 	newCol := &model.ColumnInfo{}
 	oldColName := &model.CIStr{}
-	err := job.DecodeArgs(newCol, oldColName)
+	pos := &ast.ColumnPosition{}
+	err := job.DecodeArgs(newCol, oldColName, pos)
 	if err != nil {
 		job.State = model.JobCancelled
 		return errors.Trace(err)
 	}
 
-	return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	if pos.Tp == ast.ColumnPositionNone {
+		return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	}
+
+	tblInfo, err := t.GetTable(job.SchemaID, job.TableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	originalState := tblInfo.State
+	switch tblInfo.State {
+	case model.StatePublic:
+		if err = d.modifyColumnPosition(tblInfo, oldColName, pos); err != nil {
+			job.State = model.JobCancelled
+			return errors.Trace(err)
+		}
+		// use delete only to prevent add record based on different column order.
```

Comment:
We encode the whole row in this format: col1ID, col1Val, col2ID, col2Value, col3ID, col3Value.   We chose this encode rule because we need to handle rows with different format. For example we are adding/dropping column.
See: https://github.com/pingcap/tidb/blob/master/tablecodec/tablecodec.go#L165